### DISCUSSION TITLE
29 - Contextual UI

### DISCUSSION THREAD

**Discussion by jonathanyeung** [dendronhq/dendron#1336] [RFCs/Ideas]
Discussion for Contextual UI RFC.

**Comment by SeriousBug** [comment]
This is great, I think the discoverability of the features is a big issue for Dendron right now and this is a great idea.

> Delete Note ... Seems like this already works?

Works for me, when I deleted a note I could no longer find it in the lookup. Same for creating a note, the only issue is the newly created note doesn't have a frontmatter, but we do warn against this.

> Code Errors and Warnings ... Broken Links or References ...

Good idea. This actually would be pretty easy to implement, since we already figure out which links exist to decide what needs to be highlighted.

**Comment by nickolay-kondratyev** [comment]
Makes a lot of sense to add contextual UI features. 


Feedback comments below: 
#### Context:
> This will also hopefully solve some user pain points around stale cache problems and reduce (and eventually eliminiate) the need for explicitly running the `Reload Index` command, which is not user-friendly.

Not clear how adding more commands will fix stale cache issues. 

####  Proposal: 
Very much agree that adding some contextual [Code Actions](https://code.visualstudio.com/api/references/vscode-api#CodeActionProvider) would really be great as example creating a note from [[wiki-link-file-which-hasnt-been-created-yet]] 

#### Add/Remove Vault:

> Proposed UI: `Add Vault` and `Remove Vault` options in explorer pane by the context menu - Right Click > [Add|Remove] Vault.

Do you mean adding `Add vault` and `Remove Vault` into the following menu right click? 

<img width="514" alt="Screen Shot 2021-09-20 at 12 32 02 PM" src="https://user-images.githubusercontent.com/4050134/133960468-421c68b7-2a34-4fc8-abe3-6a35283a117e.png">
If so the concern here its currently very much about just showing which panes to show.

If above is not the case and If we are adding additional buttons to the explorer we should be careful to not have too many icons stacked up. Here is what I mean: In our current UI in note look up we have quite a few icons stacked up. Including an icon to split the view for the newly created view Horizontally:

<img width="865" alt="Screen Shot 2021-09-20 at 12 40 16 PM" src="https://user-images.githubusercontent.com/4050134/133960889-62c689c7-b913-4da0-a2b7-f79c344cfb03.png">

Which only splits a single note horizontally and gets un-toggled on next note lookup. Since such functionality can be achieved by VSCode shortcut without having to click the button with a mouse, it begs the question whether such toggle should even exist in note lookup. The point being we should be careful what we decide to be top level toggles/entry points. And having some reasonable nesting for the actions we show in contextual menus could be a good idea especially as actions grow. 

## Menu
Also something we can consider for the UI additions is having a menu with Dendron commands. As example Intelij Git plugin menu for inspiration:

<img width="453" alt="Screen Shot 2021-09-20 at 12 50 34 PM" src="https://user-images.githubusercontent.com/4050134/133961448-1e8a1c0a-42c4-404d-bddb-a6fe89778b3e.png">

**Comment by kevinslin** [comment]
Agree with all the suggestions in the RFC. Can we break up this into multiple phases so we can have something ready to ship for v61?

**Comment by Harshita-mindfire** [comment]
Highlighting few points which I could discover

> Delete Note ... Seems like this already works?

Works for me as well

> Proposed

For each menu options on Right Click, there is (and should be for new options) a dedicated command that runs. Not very sure that vscode allows to change the behavior of existing options like Delete, Copy, Rename etc. 
We can either add a Dendron specific menu option like `Dendron : Delete Note` or can make use of our tree-view to have full control over menu options
